drivers/clock_control/clock_control_npcm.c

```diff
@@ -0,0 +1,365 @@
+	NPCM_PWDWN_CTL(pmc_base, NPCM_CLOCK_REG_OFFSET(priv->clock_id)) |=
+		~(BIT(NPCM_CLOCK_REG_BIT_OFFSET(priv->clock_id)));
+	return 0;
+}
+
+static int npcm_clock_control_get_subsys_rate(const struct device *dev,
+					      clock_control_subsys_t sub_system, uint32_t *rate)
+{
+	ARG_UNUSED(dev);
+	uint32_t clk_id = (uint32_t)sub_system;
+	struct clk_cfg_t *priv;
+
+	priv = npcm_get_cfg(sub_system);
+	if (!priv) {
+		LOG_ERR("Unsupported clock id %d", clk_id);
+		return -EINVAL;
+	}
+
+	switch (priv->bus) {
+	case NPCM_CLOCK_BUS_APB1:
+		*rate = NPCM_APB_CLOCK(1);
+		break;
+	case NPCM_CLOCK_BUS_APB2:
+		*rate = NPCM_APB_CLOCK(2);
+		break;
+	case NPCM_CLOCK_BUS_APB3:
+		*rate = NPCM_APB_CLOCK(3);
+		break;
+	case NPCM_CLOCK_BUS_AHB6:
+		*rate = CORE_CLK / (AHB6DIV_VAL + 1);
+		break;
+	case NPCM_CLOCK_BUS_FIU:
+		*rate = CORE_CLK / (FIUDIV_VAL + 1);
+		break;
+	case NPCM_CLOCK_BUS_I3C:
+		*rate = CORE_CLK / (I3CDIV_VAL + 1);
+		break;
+	case NPCM_CLOCK_BUS_CORE:
+		*rate = CORE_CLK;
+		break;
+	case NPCM_CLOCK_BUS_LFCLK:
+		*rate = LFCLK;
+		break;
+	case NPCM_CLOCK_BUS_FMCLK:
+		*rate = FMCLK;
+		break;
+	case NPCM_CLOCK_BUS_USB20_CLK:
+		*rate = USB20_CLK;
+		break;
+	case NPCM_CLOCK_BUS_SIO_CLK:
+		*rate = SIO_CLK;
+		break;
+	default:
+		*rate = 0U;
+		/* Invalid parameters */
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/* Clock controller driver registration */
+static struct clock_control_driver_api npcm_clock_control_api = {
+	.on = npcm_clock_control_on,
+	.off = npcm_clock_control_off,
+	.get_rate = npcm_clock_control_get_subsys_rate,
+};
+
+static int npcm_clock_control_init(const struct device *dev)
+{
+	struct cdcg_reg *const priv = (struct cdcg_reg *)(DRV_CONFIG(dev)->base_cdcg);
+	struct freq_multiplier_t *freq_p;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(freq_multiplier); i++) {
+		if (freq_multiplier[i].ofmclk == OFMCLK) {
+			freq_p = &freq_multiplier[i];
+			break;
+		}
+	}
+
+	if (i == ARRAY_SIZE(freq_multiplier)) {
+		LOG_ERR("Unsupported OFMCLK frequency %d", OFMCLK);
+		return -EINVAL;
+	}
+
+	/*
+	 * Resetting the OFMCLK (even to the same value) will make the clock
+	 * unstable for a little which can affect peripheral communication like
+	 * eSPI. Skip this if not needed.
+	 */
+	if (priv->hfcgn != freq_p->hfcgn || priv->hfcgml != freq_p->hfcgml ||
+	    priv->hfcgmh != freq_p->hfcgmh) {
+		/*
+		 * Configure frequency multiplier M/N values according to
+		 * the requested OFMCLK (Unit:Hz).
+		 */
+		priv->hfcgn = freq_p->hfcgn;
+		priv->hfcgml = freq_p->hfcgml;
+		priv->hfcgmh = freq_p->hfcgmh;
+
+		/* Load M and N values into the frequency multiplier */
+		priv->hfcgctrl |= BIT(NPCM_HFCGCTRL_LOAD);
+		/* Wait for stable */
+		while (sys_test_bit(priv->hfcgctrl, NPCM_HFCGCTRL_CLK_CHNG))
+			;
+	}
+
+	/* Set all clock prescalers of core and peripherals. */
+	priv->hfcgp = ((FPRED_VAL << 4) | AHB6DIV_VAL);
+	priv->hfcbcd = (APB1DIV_VAL | (APB2DIV_VAL << 4));
+	priv->hfcbcd1 = ((I3CDIV_VAL << 2) | FIUDIV_VAL);
```

Comment:
nit: no need for the outer parenthesis on these ones

Reply:
Thanks. Fixed in commit [81825c1](https://github.com/zephyrproject-rtos/zephyr/pull/78907/commits/81825c1e90f29449815aafc216c8cb5e5a7ac48d).